### PR TITLE
[Merged by Bors] - expose texture/image conversions as From/TryFrom

### LINKED ISSUE #2169: image texture
I need to convert image to texture. There is already a similar function in bevy_render, but it is visible in crate. Can you let go of this restriction

### PR DESCRIPTION
**Pull request by mockersf**
fixes #2169 

Instead of having custom methods with reduced visibility, implement `From<image::DynamicImage> for Texture` and `TryFrom<Texture> for image::DynamicImage`

### CODE REVIEW

File: crates/bevy_render/src/texture/image_texture_conversion.rs [hunk 1]
```diff
@@ -1,152 +1,158 @@
...
     }
+}
 
-    Texture::new(
-        Extent3d::new(width, height, 1),
-        TextureDimension::D2,
-        data,
-        format,
-    )
+pub enum TextureConversionError {
```
Comment (NathanSWard) [inline crates/bevy_render/src/texture/image_texture_conversion.rs:125]:
```suggestion
#[derive(Clone, Copy, Debug)] // also maybe Eq/PartialEq
pub enum TextureConversionError {
```

Also is it worth implementing `std::error::Error` for this?

File: crates/bevy_render/src/texture/image_texture_conversion.rs [hunk 2]
```diff
@@ -1,152 +1,158 @@
...
+                let mut local_data =
+                    Vec::with_capacity(width as usize * height as usize * format.pixel_size());
 
-            let raw_data = i.into_raw();
+                for pixel in image.into_raw().chunks_exact(3) {
+                    // TODO unsafe_get in release builds?
+                    let r = pixel[0];
+                    let g = pixel[1];
+                    let b = pixel[2];
+                    let a = u16::max_value();
```
Comment (NathanSWard) [inline crates/bevy_render/src/texture/image_texture_conversion.rs:95]:
```suggestion
                    let a = u16::MAX;
```
`max_value()` is deprecated in favor of `::MAX`

File: crates/bevy_render/src/texture/image_texture_conversion.rs [hunk 3]
```diff
@@ -1,152 +1,161 @@
...
-    )
+        Texture::new(
+            Extent3d::new(width, height, 1),
+            TextureDimension::D2,
+            data,
+            format,
+        )
+    }
+}
+#[derive(Clone, Copy, Debug, Eq, Error, PartialEq)]
```
Comment (NathanSWard) [inline crates/bevy_render/src/texture/image_texture_conversion.rs:125]:
```suggestion
}

#[derive(Clone, Copy, Debug, Eq, Error, PartialEq)]
```

**Review by NathanSWard — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/texture/image_texture_conversion.rs (hunk 3 above)
Comment (NathanSWard) [reply crates/bevy_render/src/texture/image_texture_conversion.rs:125]: Missing a newline here ;)

**Review by NathanSWard — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
Love it!

**Pull request closed without merging**